File: android/src/main/kotlin/com/retozu/flutter_amazon_chime/DataMessageObserver.kt
package com.retozu.flutter_amazon_chime

import com.amazonaws.services.chime.sdk.meetings.realtime.datamessage.DataMessage
import com.amazonaws.services.chime.sdk.meetings.realtime.datamessage.DataMessageObserver

class ChimeDataMessageObserver(private val chimeFlutterApi: ChimeFlutterApi) : DataMessageObserver {
    /** The wildcard topic used to receive all data messages. */
    val topic = "*"

    override fun onDataMessageReceived(dataMessage: DataMessage) {
        val text = dataMessage.text() ?: ""
        val msg = DataMessageMsg(
            topic = dataMessage.topic,
            data = text,
            senderAttendeeId = dataMessage.senderAttendeeId,
            senderExternalUserId = dataMessage.senderExternalUserId,
            timestampMs = dataMessage.timestampMs,
            throttled = dataMessage.throttled
        )
        chimeFlutterApi.onDataMessageReceived(msg) {}
    }
}

File: ios/Classes/DataMessageObserver.swift
import AmazonChimeSDK

class ChimeDataMessageObserver: DataMessageObserver {
    let chimeFlutterApi: ChimeFlutterApi
    /// The wildcard topic used to receive all data messages.
    let topic = "*"

    init(chimeFlutterApi: ChimeFlutterApi) {
        self.chimeFlutterApi = chimeFlutterApi
    }

    func dataMessageDidReceived(dataMessage: DataMessage) {
        let text = String(data: dataMessage.data, encoding: .utf8) ?? ""
        let msg = DataMessageMsg(
            topic: dataMessage.topic,
            data: text,
            senderAttendeeId: dataMessage.senderAttendeeId,
            senderExternalUserId: dataMessage.senderExternalUserId,
            timestampMs: dataMessage.timestampMs,
            throttled: dataMessage.throttled
        )
        chimeFlutterApi.onDataMessageReceived(message: msg) { _ in }
    }
}
